misc/jobserver_pool.py

```diff
@@ -0,0 +1,297 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Setup a GNU Make Jobserver token pool then launch a command with it.
+
+On Windows, this only supports the semaphore-based scheme.
+On Posix, this uses a FIFO by default, except if `--pipe` is used.
+
+NOTE: This is a basic implementation that doesn't support broken
+      protocol clients that release more tokens than they acquired
+      to the pool. Using these in your build will result in extra job
+      slots being created, severely degrading overall performance
+      over time.
+
+See --help-usage for usage examples.
+"""
+import argparse
+import os
+import platform
+import subprocess
+import sys
+import typing as T
+
+_DEFAULT_NAME = "jobserver_tokens"
+_IS_WINDOWS = sys.platform in ("win32", "cygwin")
+
+if _IS_WINDOWS:
+    try:
+        # This requires pywin32 to be installed.
+        import win32event
+        import win32api
+        import winerror
+    except ModuleNotFoundError as e:
+        print(
+            "\nERROR: Could not import Win32 API, please install pywin32, e.g. `python -m pip install pywin32`.\n",
+            file=sys.stderr,
+        )
+        raise e
+
+    def create_sem(
+        sem_name: str, token_count: int
+    ) -> T.Tuple[win32event.PyHandle, T.Dict[str, str]]:
+        """Create and initialize Win32 semaphore."""
+        assert token_count > 0, f"Token count must be strictly positive"
+        handle = win32event.CreateSemaphore(
+            None, token_count, token_count - 1, sem_name  # Default security attributes,
+        )
+        assert handle != 0, f"Error creating Win32 semaphore {winerror.GetLastError()}"
+        env = dict(os.environ)
+        env["MAKEFLAGS"] = " -j{token_count} --jobserver-auth=" + sem_name
+        return handle, env
+
+    def check_sem_count(handle: win32event.PyHANDLE, token_count: int) -> int:
+        if token_count <= 1:
+            # Nothing to check here.
+            return 0
+
+        expected_count = token_count - 1
+
+        read_count = win32event.ReleaseSemaphore(handle, 1)
+        if read_count < expected_count:
+            print(
+                f"ERROR: {expected_count - read_count} were missing from the token pool (got {read_count}, expected {expected_count})",
+                file=sys.stderr,
+            )
+            return 1
+        if read_count > expected_count:
+            print(
+                f"ERROR: {read_count - expected_count} extra tokens were released to the token pool (got {read_count}, expected {expected_count})",
+                file=sys.stderr,
+            )
+            return 1
+
+        return 0
+
+    def print_usage() -> int:
+        print(
+            """Example usage:
+
+# Start <command> after setting the server to provide as many tokens
+# as available CPUs (the default)
+python \\path\\to\\jobserver_pool.py <command>
```

Comment:
Forward slashes also works on Windows so you can make one single help text.

Reply:
I prefer to keep separate texts and follow each platform's favorite idions here.

We need separate texts for Windows and Posix because the script doesn't accept the same set of options on them. Besides, the examples with `bash` and `powershell.exe` are also different.

Another issue is that on Windows you *need* to put `python` to call the Python3 interpreter, otherwise nothing happens (on Powershell), or you get an error (with cmd.exe).

While on Posix you can invoke the script directly, and you would need `python3`, not `python` which on Linux is usually non-existent, or tied to `python2`, and in rare cases an alias for `python3`.

Reply:
I think people developing ninja knows how to start their python interpreter. Having the text duplicated makes it harder to read the usage instructions by reading the source code. If you still insist on having backward slashes, please use r""" to not having to escape backslashes.